src/backend/distributed/metadata/node_metadata.c

```diff
@@ -753,13 +754,662 @@ PgDistTableMetadataSyncCommandList(void)
+	metadataDeletionCommands = lappend(metadataDeletionCommands,
+									   DELETE_ALL_PARTITIONS);
+	metadataDeletionCommands = lappend(metadataDeletionCommands,
+									   DELETE_ALL_SHARDS);
+	metadataDeletionCommands = lappend(metadataDeletionCommands,
+									   DELETE_ALL_PLACEMENTS);
+	metadataDeletionCommands = lappend(metadataDeletionCommands,
+									   DELETE_ALL_DISTRIBUTED_OBJECTS);
+	metadataDeletionCommands = lappend(metadataDeletionCommands,
+									   DELETE_ALL_COLOCATION);
+	metadataDeletionCommands = lcons(DISABLE_DDL_PROPAGATION,
+									 metadataDeletionCommands);
+	metadataDeletionCommands = lappend(metadataDeletionCommands,
+									   ENABLE_DDL_PROPAGATION);
+	SendMetadataCommandListToWorkerListInCoordinatedTransaction(nodesWithMetadata,
+																CurrentUserName(),
+																metadataDeletionCommands);
+	list_free(nodesWithMetadata);
+}
+
+
+/*
+ * SendIterableTableMetadataCommands sends all tables' metadata within metadata
+ * connection.
+ */
+void
+SendIterableTableMetadataCommands(List *nodeList)
+{
+	/* find relevant worker nodes */
+	List *nodesWithMetadata = NIL;
+	WorkerNode *workerNode = NULL;
+	foreach_ptr(workerNode, nodeList)
+	{
+		if (NodeIsPrimary(workerNode) && !NodeIsCoordinator(workerNode))
+		{
+			nodesWithMetadata = lappend(nodesWithMetadata, workerNode);
+		}
+	}
+
+	if (nodesWithMetadata == NIL)
+	{
+		return;
+	}
+
+	ScanKeyData scanKey[1];
+	int scanKeyCount = 0;
+
+	Relation relation = table_open(DistPartitionRelationId(), AccessShareLock);
+
+	TupleDesc tupleDesc = RelationGetDescr(relation);
+	SysScanDesc scanDesc = systable_beginscan(relation,
+											  InvalidOid, false,
+											  NULL, scanKeyCount, scanKey);
+
+	MemoryContext metadataSyncContext = AllocSetContextCreate(CurrentMemoryContext,
+															  "metadata sync context",
+															  ALLOCSET_DEFAULT_SIZES);
+	MemoryContext oldContext = MemoryContextSwitchTo(metadataSyncContext);
+
+	bool scanEnded = false;
+	HeapTuple nextTuple = NULL;
+	while (!scanEnded)
+	{
+		MemoryContextReset(metadataSyncContext);
+
+		MemoryContextSwitchTo(oldContext);
+		nextTuple = systable_getnext(scanDesc);
+		if (!HeapTupleIsValid(nextTuple))
+		{
+			scanEnded = true;
+			MemoryContextSwitchTo(metadataSyncContext);
+			continue;
+		}
+		else if (!ShouldMetadataSyncHeapTuple(nextTuple, tupleDesc))
+		{
+			MemoryContextSwitchTo(metadataSyncContext);
+			continue;
+		}
+		MemoryContextSwitchTo(metadataSyncContext);
+
+		if (nextTuple)
+		{
+			bool isNullArray[Natts_pg_dist_partition];
+			Datum datumArray[Natts_pg_dist_partition];
+			heap_deform_tuple(nextTuple, tupleDesc, datumArray,
+							  isNullArray);
+			Datum relationIdDatum = datumArray[Anum_pg_dist_partition_logicalrelid - 1];
+			Oid relationId = DatumGetObjectId(relationIdDatum);
+
+			List *tableMetadataCreateCommandList =
+				CitusTableMetadataCreateCommandList(relationId);
+			tableMetadataCreateCommandList = lcons(DISABLE_DDL_PROPAGATION,
```

Comment:
do we really need lcons/lappend? Those are not always cheap. In other words, can't we just do:
```C
SendMetadataCommandListToWorkerListInCoordinatedTransaction(DISABLE_DDL_PROPAGATION);
SendMetadataCommandListToWorkerListInCoordinatedTransaction(tableMetadataCreateCommandList);
SendMetadataCommandListToWorkerListInCoordinatedTransaction(ENABLE_DDL_PROPAGATION);

Reply:
We reset memory context after we send commands for each object, so they should always be cheap in our case.